# Phase 1 — Inspect the Raw Data

###Step 1 — Import Libraries

In [1]:
import pandas as pd
import numpy as np

Pandas is a Python library used for working with data sets. It has functions for analyzing, cleaning, exploring, and manipulating data.<br/>
pd is just an alias for Pandas. Using it makes the code shorter when calling methods or properties.<br/>


NumPy is a Python library used for working with arrays.
It also has functions for working in domain of linear algebra, fourier transform, and matrices.<br/>
NumPy stands for Numerical Python.<br/>
np is the standard alias for the NumPy library.


This cell imports the required Python libraries for the project. Pandas (pd) is used for data manipulation and analysis, while NumPy (np) is used for numerical and mathematical operations.

###Step 2 — Load Dataset

In [2]:
df = pd.read_csv("/content/drive/MyDrive/employee_attendance_bangalore/employee_attendance_bangalore_q1_2026.csv")

df stands for DataFrame.
A DataFrame is a two-dimensional table (rows and columns) provided by the Pandas library.
We use df as a variable name to store the dataset after it is loaded.

pd is the alias (short name) for the Pandas library.
read_csv() is a Pandas function used to read a CSV (Comma-Separated Values) file and load it into a DataFrame.

Google Drive is mounted to allow Google Colab to access files stored in the user's Drive. The dataset is then loaded from the specified file path using pd.read_csv() and stored in the DataFrame df.

###Step 3 — Display First 5 Rows

In [3]:
df.head()

,attendance_id,employee_id,employee_name,gender,department,designation,employment_type,office_location,date_of_joining,attendance_date,...,work_mode,login_timestamp,logout_timestamp,total_hours_worked,break_duration_mins,net_productive_hours,late_arrival_mins,early_exit_mins,overtime_hours,leave_type
0,ATT0000001,VL1000,Arjun Verma,Male,Engineering,Software Engineer Intern,Intern,Koramangala HQ,2025-06-26,2026-01-02,...,Work From Office,2026-01-02 09:42:03,2026-01-02 18:34:12,8.87,86,7.44,12,0,0.00,Not Applicable
1,ATT0000002,VL1001,Rekha Chatterjee,Female,Operations,Operations Manager,Full-Time,Koramangala HQ,2024-02-16,2026-01-02,...,Work From Home,2026-01-02 09:09:11,2026-01-02 18:25:11,9.27,67,8.15,0,6,0.00,Not Applicable
2,ATT0000003,VL1002,Tanmay Bhardwaj,Male,Operations,Operations Analyst,Full-Time,Whitefield Tech Park,2020-10-13,2026-01-02,...,Work From Home,2026-01-02 09:45:59,2026-01-02 20:17:31,10.53,57,9.58,15,0,0.58,Not Applicable
3,ATT0000004,VL1003,Aditya Bose,Female,Customer Success,Customer Success Executive,Full-Time,Koramangala HQ,2022-01-10,2026-01-02,...,Work From Office,2026-01-02 13:49:26,2026-01-02 22:24:51,8.59,76,7.32,0,36,0.00,Not Applicable
4,ATT0000005,VL1004,Uday Acharya,Male,Human Resources,Talent Acquisition Partner,Full-Time,Electronic City Campus,2024-09-09,2026-01-02,...,Work From Office,2026-01-02 10:51:02,2026-01-02 20:33:39,9.71,61,8.69,20,0,0.00,Not Applicable


df is the DataFrame that contains the dataset.<br/>
head() is a Pandas function that displays the first 5 rows of the DataFrame.<br/>
df.head() displays the first 5 rows of the dataset.

###Step 4 — Dataset Shape

In [4]:
print("Rows :", df.shape[0])
print("Columns :", df.shape[1])

Rows : 55374
Columns : 22


print() is a built-in Python function used to display output on the screen.<br/>
shape is a Pandas DataFrame attribute that returns the dimensions of the dataset as a tuple:(number_of_rows, number_of_columns)<br/>
For this dataset the tuple : (55374,22)<br/>
df.shape--> should return (55374,22)<br/>
Tuple indexing is applied here<br/>
df.shape[0]   → 5000 (Rows)<br/>
df.shape[1]   →   12 (Columns)

###Step 5 — Column Names

In [5]:
df.columns

Index(['attendance_id', 'employee_id', 'employee_name', 'gender', 'department',
       'designation', 'employment_type', 'office_location', 'date_of_joining',
       'attendance_date', 'shift_type', 'attendance_status', 'work_mode',
       'login_timestamp', 'logout_timestamp', 'total_hours_worked',
       'break_duration_mins', 'net_productive_hours', 'late_arrival_mins',
       'early_exit_mins', 'overtime_hours', 'leave_type'],
      dtype='object')

df.columns used to view all the column names in the dataset.

###Step 6 — Dataset Information

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55374 entries, 0 to 55373
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   attendance_id         55374 non-null  object 
 1   employee_id           55374 non-null  object 
 2   employee_name         55374 non-null  object 
 3   gender                55374 non-null  object 
 4   department            55374 non-null  object 
 5   designation           55374 non-null  object 
 6   employment_type       55374 non-null  object 
 7   office_location       55374 non-null  object 
 8   date_of_joining       55374 non-null  object 
 9   attendance_date       55374 non-null  object 
 10  shift_type            55374 non-null  object 
 11  attendance_status     55374 non-null  object 
 12  work_mode             55374 non-null  object 
 13  login_timestamp       52739 non-null  object 
 14  logout_timestamp      52739 non-null  object 
 15  total_hours_worked 

df.info() is one of the most useful Pandas functions for understanding a dataset.

###Step 7 — Basic Statistics

In [7]:
df.describe()

,total_hours_worked,break_duration_mins,net_productive_hours,late_arrival_mins,early_exit_mins,overtime_hours
count,55374.000000,55374.000000,55374.000000,55374.000000,55374.000000,55374.000000
mean,8.724951,59.741955,7.729238,9.858471,20.099198,0.150049
std,2.413434,23.878616,2.223308,19.465159,49.630750,0.367258
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,8.470000,47.000000,7.370000,0.000000,0.000000,0.000000
50%,9.280000,62.000000,8.210000,0.000000,0.000000,0.000000
75%,10.000000,76.000000,8.960000,14.000000,16.000000,0.000000
max,13.660000,135.000000,12.860000,218.000000,391.000000,3.860000


**df.describe()** provides a statistical summary of the numerical columns in a DataFrame. It helps analyze the data by displaying count, mean, standard deviation, minimum, maximum, and quartile values, making it useful for understanding data distribution and identifying potential outliers.

**df.describe()** helps to understand the **spread of the data** and tells the **range** in which the data is distributed.

Explain each row:<br/>
1. **Count** : <br/>
Each numerical column contains 55,374 records, so there are no missing values in these numerical columns.<br/>

2. **Mean** :
- On average, employees worked 8.72 hours every day.
- Employees took approximately 60 minutes of break on average.
- Employees were productive for nearly 7.7 hours on average.
- Employees arrived approximately 10 minutes late on average.
- Employees left about 20 minutes early on average.
- Employees performed very little overtime on average.

3. **Standard Deviation (std)** :The Standard Deviation (std) row tells us how much the values in each column vary from the average (mean).
- Employees' working hours vary by about 2.4 hours from the average.
- Break durations differ by about 24 minutes from the average break time (59.74 minutes). Some employees take much shorter breaks, while others take much longer ones.
- Productive working hours vary by approximately 2.22 hours from the average (7.73 hours). Most employees' productive time is relatively consistent.
- Late arrival times vary by around 19.47 minutes from the average (9.86 minutes). This suggests that while many employees arrive on time, some arrive significantly late, increasing the variation.
- Early exit times have a high standard deviation of about 49.63 minutes, meaning there is a large difference in employees' early leaving behavior. Some employees never leave early, while a few leave much earlier than usual.
- Overtime hours vary by only 0.37 hours (about 22 minutes) from the average (0.15 hours). This indicates that most employees do not perform much overtime, and overtime is generally low.

4. **Minimum** :
- At least one employee worked 0 hours on a particular day. This could indicate absence, leave, a holiday, or that the employee did not log any work hours.
- At least one employee took no break during the workday. This may happen if the employee worked for a short duration, forgot to record a break, or genuinely did not take one.
- At least one employee had 0 productive hours. This usually corresponds to an absent employee or someone who did not perform any productive work that day.
- At least one employee was not late at all. A value of 0 means the employee arrived on time or earlier than the scheduled reporting time.
- At least one employee did not leave early. They either completed their scheduled shift or stayed until the expected logout time.
- At least one employee (and likely many employees) did not work any overtime beyond their regular working hours.

5. **25% or First Quartile (Q1)** : 25% of the data values are less than or equal to this value, and the remaining 75% are greater than this value.
- The first quartile for total working hours is 8.47 hours, indicating that one-fourth(25%) of the employees worked 8.47 hours or less, while the remaining(75%) employees worked longer.
- The first quartile of break duration is 47 minutes, meaning that 25% of employees took relatively shorter breaks.
- The first quartile of productive hours is 7.37 hours, indicating that only one-fourth of employees had productive time below this level.
- The first quartile of late arrival is 0 minutes, which shows that at least 25% of employees had no late arrivals.
- The first quartile of early exit is 0, indicating that at least one-fourth of employees completed their full working hours without leaving early.
- The first quartile of overtime is 0 hours, meaning that at least 25% of employees worked only their regular hours.

6. **50% or Median or Second Quartile (Q2)** : The Median is the middle value when all the data is arranged in ascending order.
- The median working time is 9.28 hours, indicating that 50% of employees worked up to 9.28 hours, while the remaining 50% worked longer.
- The median break duration is 62 minutes, showing that half of the employees took breaks of one hour or less.
- The median productive time is 8.21 hours, which indicates that a typical employee was productive for about 8 hours during the workday.
- The median late arrival is 0 minutes, meaning that at least half of the employees arrived on time.
- The median early exit is 0 minutes, indicating that at least 50% of employees completed their full working hours.
- The median overtime is 0 hours, which suggests that overtime was not common for the majority of employees.

7. **75% or Third Quartile (Q3)** : 75% of the observations are less than or equal to this value, and only the remaining 25% are greater than this value.
- The third quartile of total working hours is 10 hours, indicating that most (75%) employees worked 10 hours or less, while only a small proportion (25%) worked beyond 10 hours.
- The third quartile for break_duration_mins is 76 minutes. This indicates that 75% of employees completed their breaks within 76 minutes, while only 25% took longer breaks.
- The third quartile for net_productive_hours is 8.96 hours, meaning that 75% of employees were productive for 8.96 hours or less, while only 25% exceeded this value.
- The third quartile for late_arrival_mins is 14 minutes, indicating that 75% of employees were late by no more than 14 minutes, while only 25% were late by more than 14 minutes.
- The third quartile for early_exit_mins is 16 minutes, meaning that 75% of employees left at most 16 minutes early, while 25% left earlier than that.
- The third quartile for overtime_hours is 0 hours, which means 75% of employees did not work any overtime at all. Only the remaining 25% recorded overtime.

8. **max** : The maximum (max) value is the highest observation recorded in that column.
- The maximum value of total_hours_worked is 13.66 hours, which means the employee who worked the longest recorded 13.66 working hours in a single day.
- The maximum break duration is 135 minutes, meaning at least one employee took a break lasting 2 hours and 15 minutes.
- The maximum value of net_productive_hours is 12.86 hours, indicating that one employee remained productive for nearly 13 hours.
- The maximum late arrival is 218 minutes, meaning one employee arrived approximately 3 hours and 38 minutes late.
- The maximum early exit is 391 minutes, indicating that one employee left approximately 6 hours and 31 minutes before the scheduled end of the workday.
- The maximum overtime is 3.86 hours, meaning one employee worked almost 4 hours beyond the regular working schedule.



###Step 8 — Missing Values

In [8]:
df.isnull().sum()

,0
attendance_id,0
employee_id,0
employee_name,0
gender,0
department,0
designation,0
employment_type,0
office_location,0
date_of_joining,0
attendance_date,0


The reason why we are not cleaning the missing values in login_timestamp and logout_timestamp:<br/>
- login_timestamp	2635
- logout_timestamp	2635<br/>
- Both the values are same.<br/>

Possible reasons include:

- Employee was absent.
- Employee was on leave.
- It was a holiday.
- No attendance was recorded for that day.

If an employee never came to the office, there would naturally be:

- no login time,
- no logout time.

These missing values are **meaningful**, not necessarily errors.

###Step 9 — Duplicate Rows

In [9]:
df.duplicated().sum()

np.int64(0)

###Step 10 — Duplicate attendance_id

In [10]:
df["attendance_id"].duplicated().sum()

np.int64(0)

attendance_id should not repeat as each employee should have different attendance_id in different days. So attendance_id is not having any duplicate value.

###Step 11 — Duplicate employee_id

In [11]:
df["employee_id"].duplicated().sum()

np.int64(54394)

employee_id is expected to repeat because each employee attends multiple days with different attendance_id, but the employee_id remains same. So employee_id is having duplicate values.

###Step 12 — Data Types

**df.dtypes** shows the data-type of each column in the dataset

In [12]:
df.dtypes

,0
attendance_id,object
employee_id,object
employee_name,object
gender,object
department,object
designation,object
employment_type,object
office_location,object
date_of_joining,object
attendance_date,object


###Step 13 — Memory Usage

It tells us how much RAM each column uses while the DataFrame is loaded in Python.

In [13]:
df.memory_usage(deep=True)

,0
Index,132
attendance_id,3267066
employee_id,3045570
employee_name,3442982
gender,2980742
department,3312109
designation,3830839
employment_type,3198638
office_location,3651166
date_of_joining,3267066


- df : dataframe
- .memory_usage() : Pandas method


In Python, **text columns** like **employee_name** or **department** are stored as **string objects**. Instead of storing the complete text directly inside each DataFrame cell, Pandas stores a **reference (or pointer)** to where that **text is kept in memory**. By default, **df.memory_usage()** only counts the **memory used by these references**. When we use **deep=True**, Pandas also counts the **memory used by the actual text values**. This gives us the real and more accurate memory usage of the DataFrame.

###Step 14 — Unique Values

In [14]:
for col in df.columns:
    print(col, ":", df[col].nunique())

attendance_id : 55374
employee_id : 980
employee_name : 980
gender : 3
department : 10
designation : 48
employment_type : 4
office_location : 5
date_of_joining : 804
attendance_date : 57
shift_type : 5
attendance_status : 3
work_mode : 4
login_timestamp : 50870
logout_timestamp : 51738
total_hours_worked : 1027
break_duration_mins : 135
net_productive_hours : 1021
late_arrival_mins : 188
early_exit_mins : 368
overtime_hours : 287
leave_type : 9


It loops through every column in the DataFrame and prints the number of unique  values present in that column.

# Phase 2a: Data Cleaning & Preparation

Make a Copy of the original dataset

In [15]:
clean_df = df.copy()

Display all the column names so that we can inspect them.

In [16]:
clean_df.columns

Index(['attendance_id', 'employee_id', 'employee_name', 'gender', 'department',
       'designation', 'employment_type', 'office_location', 'date_of_joining',
       'attendance_date', 'shift_type', 'attendance_status', 'work_mode',
       'login_timestamp', 'logout_timestamp', 'total_hours_worked',
       'break_duration_mins', 'net_productive_hours', 'late_arrival_mins',
       'early_exit_mins', 'overtime_hours', 'leave_type'],
      dtype='object')

###Step 1 : Remove extra spaces from column headers

In [17]:
clean_df.columns = clean_df.columns.str.strip()

It removes any extra spaces at the beginning (leading) and end (trailing) of all column headers in the DataFrame.

###Step 2 — Remove Extra Spaces from Text Columns

Storing text column names in the variable :

In [18]:
text_columns = clean_df.select_dtypes(include="object").columns

It finds all the text (string/object) columns in the DataFrame and store their names in the variable text_columns.

- clean_df : dataframe
- .select_dtypes() : Pandas method, selects columns based on their data type
- include="object" : Select only the columns whose data type is object
- .columns : it returns only the selected column names

Now clean them:

In [19]:
for col in text_columns:
    clean_df[col] = clean_df[col].str.strip()

The loop automatically goes through:
all the columns like
- employee_name
- gender
- department

and removes extra spaces from the values in each of those columns.

###Step 3 — Convert Date Columns

Many date and time columns are initially stored as text (object/string) when a CSV file is imported into Pandas. To perform date-based analysis, these columns must be converted into the datetime data type.

In [20]:
date_columns = [
    "date_of_joining",
    "attendance_date",
    "login_timestamp",
    "logout_timestamp"
]

for col in date_columns:
    clean_df[col] = pd.to_datetime(clean_df[col], errors="coerce")

A list named date_columns is created to store the names of all the date and time columns. Each column in the list is then processed one by one using a for loop. The **pd.to_datetime()** function is used to convert each column from the string (object) data type to the datetime data type. The **errors='coerce'** parameter is used to ensure that any invalid or incorrectly formatted date values are converted to NaT (Not a Time) instead of causing an error. After the conversion, various date-based operations such as extracting the year, month, day, and calculating time differences can be performed.

In [21]:
clean_df["attendance_date"].dt.month

,attendance_date
0,1
1,1
2,1
3,1
4,1
...,...
55369,3
55370,3
55371,3
55372,3


This line extracts the month from every value in the attendance_date column.
<br/>
- clean_df : DataFrame
- clean_df["attendance_date"] : This selects the attendance_date column from the DataFrame.
- .dt is a Pandas datetime accessor. It allows to work with datetime columns and extract different parts of a date.
- .month extracts the month from each date.

###Step 4 — Verify Missing Login/Logout

- To find the reason behind the missing login_timestamp values.
- To check whether the missing values are actual missing data or valid leave records.

In [22]:
clean_df[
    clean_df["login_timestamp"].isnull()
][
    ["attendance_status", "leave_type"]
].head(10)

,attendance_status,leave_type
15,On Leave,Loss of Pay
18,On Leave,Loss of Pay
23,On Leave,Sick Leave
69,On Leave,Sick Leave
74,On Leave,Casual Leave
82,On Leave,Sick Leave
91,On Leave,Casual Leave
150,On Leave,Casual Leave
159,On Leave,Casual Leave
184,On Leave,Casual Leave


- clean_df["login_timestamp"].isnull() checks which rows have missing login timestamps.
- Only those rows are selected.
- From those rows, only the attendance_status and leave_type columns are displayed.
- .head(10) shows the first 10 matching records.

###Step 5 — Verify IDs

In [23]:
print("Duplicate attendance_id:",
      clean_df["attendance_id"].duplicated().sum())

print("Duplicate employee_id:",
      clean_df["employee_id"].duplicated().sum())

Duplicate attendance_id: 0
Duplicate employee_id: 54394


# Phase 2b: Creating new column

###Step 1: Create Joining_Year

In [24]:
clean_df["Joining_Year"] = clean_df["date_of_joining"].dt.year

This line extracts the year from the date_of_joining column and creates a new column called Joining_Year.

- clean_df : DataFrame
- ["date_of_joining"] : This selects the date_of_joining column.
- .dt : datetime accessor allows us to extract different parts of a date
- .year : extracts only the year from each date.

Verify the result:

In [25]:
clean_df[["date_of_joining", "Joining_Year"]].head()

,date_of_joining,Joining_Year
0,2025-06-26,2025
1,2024-02-16,2024
2,2020-10-13,2020
3,2022-01-10,2022
4,2024-09-09,2024


This line displays the first 5 rows of the **date_of_joining** and **Joining_Year** columns to verify that the Joining_Year column has been created correctly.

###Step 2: Create Employee_Tenure_Years

- To calculate how many years each employee has been working in the company.
- To create a new column called Employee_Tenure_Years for employee tenure analysis.

- The joining year is already stored in the Joining_Year column.
- The attendance year is extracted from the attendance_date column using .dt.year.
- The employee's tenure is calculated by subtracting the joining year from the attendance year.

- Employee Tenure (Years) = Attendance Year − Joining Year

In [26]:
clean_df["Employee_Tenure_Years"] = (
    clean_df["attendance_date"].dt.year -
    clean_df["Joining_Year"]
)

- To verify that the Employee_Tenure_Years column has been calculated correctly.

In [27]:
clean_df[
    ["Joining_Year", "attendance_date", "Employee_Tenure_Years"]
].head()

,Joining_Year,attendance_date,Employee_Tenure_Years
0,2025,2026-01-02,1
1,2024,2026-01-02,2
2,2020,2026-01-02,6
3,2022,2026-01-02,4
4,2024,2026-01-02,2


- The Joining_Year, attendance_date, and Employee_Tenure_Years columns are selected.
- The .head() method displays the first 5 records.

###Step 3: Create Attendance_Month

- To extract the month name from the attendance_date column.
- To create a new column named Attendance_Month for month-wise analysis and reporting.

In [28]:
clean_df["Attendance_Month"] = clean_df["attendance_date"].dt.month_name()

- The attendance_date column is already in the datetime format.
- The .dt.month_name() function extracts the full month name from each date.
- The extracted month name is stored in a new column called Attendance_Month.

Verify:

In [29]:
clean_df[
    ["attendance_date", "Attendance_Month"]
].head()

,attendance_date,Attendance_Month
0,2026-01-02,January
1,2026-01-02,January
2,2026-01-02,January
3,2026-01-02,January
4,2026-01-02,January


- clean_df : DataFrame
- [["attendance_date", "Attendance_Month"]] : This selects only these two columns from the DataFrame.

 - attendance_date → Original attendance date.
 - Attendance_Month → Month name extracted from the attendance date.
- Double square bracket are used as multiple columns are being selected.
- .head() method displays the first 5 rows of the selected columns.

- To verify that the Attendance_Month column has been created correctly.
- To compare the original attendance_date with the extracted month name.

###checking the new dataset once:

In [30]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55374 entries, 0 to 55373
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   attendance_id          55374 non-null  object        
 1   employee_id            55374 non-null  object        
 2   employee_name          55374 non-null  object        
 3   gender                 55374 non-null  object        
 4   department             55374 non-null  object        
 5   designation            55374 non-null  object        
 6   employment_type        55374 non-null  object        
 7   office_location        55374 non-null  object        
 8   date_of_joining        55374 non-null  datetime64[ns]
 9   attendance_date        55374 non-null  datetime64[ns]
 10  shift_type             55374 non-null  object        
 11  attendance_status      55374 non-null  object        
 12  work_mode              55374 non-null  object        
 13  l

- Originally: 22 columns
- 22 + 3(Joining_Year,Employee_Tenure_Years, Attendance_Month) = 25 columns



Exactly as expected.

In [31]:
clean_df.head()

,attendance_id,employee_id,employee_name,gender,department,designation,employment_type,office_location,date_of_joining,attendance_date,...,total_hours_worked,break_duration_mins,net_productive_hours,late_arrival_mins,early_exit_mins,overtime_hours,leave_type,Joining_Year,Employee_Tenure_Years,Attendance_Month
0,ATT0000001,VL1000,Arjun Verma,Male,Engineering,Software Engineer Intern,Intern,Koramangala HQ,2025-06-26,2026-01-02,...,8.87,86,7.44,12,0,0.00,Not Applicable,2025,1,January
1,ATT0000002,VL1001,Rekha Chatterjee,Female,Operations,Operations Manager,Full-Time,Koramangala HQ,2024-02-16,2026-01-02,...,9.27,67,8.15,0,6,0.00,Not Applicable,2024,2,January
2,ATT0000003,VL1002,Tanmay Bhardwaj,Male,Operations,Operations Analyst,Full-Time,Whitefield Tech Park,2020-10-13,2026-01-02,...,10.53,57,9.58,15,0,0.58,Not Applicable,2020,6,January
3,ATT0000004,VL1003,Aditya Bose,Female,Customer Success,Customer Success Executive,Full-Time,Koramangala HQ,2022-01-10,2026-01-02,...,8.59,76,7.32,0,36,0.00,Not Applicable,2022,4,January
4,ATT0000005,VL1004,Uday Acharya,Male,Human Resources,Talent Acquisition Partner,Full-Time,Electronic City Campus,2024-09-09,2026-01-02,...,9.71,61,8.69,20,0,0.00,Not Applicable,2024,2,January


###Step 4 — Create Attendance_Weekday

It tells us on which day of the week is employee attendance highest

In [32]:
clean_df["Attendance_Weekday"] = clean_df["attendance_date"].dt.day_name()

- clean_df : DataFrame
- ["Attendance_Weekday"] : This creates a new column named Attendance_Weekday.
The extracted day names will be stored in this column.
- clean_df["attendance_date"].dt.day_name() : This expression selects the attendance_date column, uses the .dt accessor to work with datetime values, and applies the **.day_name()** method to extract the weekday name (such as Monday, Tuesday, or Friday) for each date. It returns the weekday names but does not create or modify any column by itself.


Verify

In [33]:
clean_df[["attendance_date", "Attendance_Weekday"]].head()

,attendance_date,Attendance_Weekday
0,2026-01-02,Friday
1,2026-01-02,Friday
2,2026-01-02,Friday
3,2026-01-02,Friday
4,2026-01-02,Friday


This code displays the first 5 rows of the attendance_date and Attendance_Weekday columns to verify that the Attendance_Weekday column has been created correctly.

In [34]:
clean_df["Attendance_Weekday"].value_counts()

,count
Attendance_Weekday,
Tuesday,12636
Friday,11646
Monday,10692
Thursday,10681
Wednesday,9719


It prints the number of times each weekday appears in the Attendance_Weekday column.

Let's verify the total:
12636 + 11646 + 10692 + 10681 + 9719 = 55374

This equals the total number of records, so your Attendance_Weekday column has been created correctly.

###Step 5 — Login_Hour

In [35]:
clean_df["Login_Hour"] = clean_df["login_timestamp"].dt.hour

This line extracts the hour from the login_timestamp column and creates a new column called Login_Hour.

Verify:

In [36]:
clean_df[["login_timestamp", "Login_Hour"]].sample(10)

,login_timestamp,Login_Hour
33552,2026-02-24 10:55:07,10.0
39201,2026-03-05 10:58:36,10.0
7359,2026-01-13 14:16:58,14.0
38794,2026-03-05 14:21:29,14.0
25489,2026-02-11 10:18:38,10.0
39095,2026-03-05 08:21:23,8.0
4017,NaT,NaN
29105,2026-02-18 09:31:11,9.0
11818,2026-01-21 09:44:18,9.0
47397,2026-03-17 10:36:02,10.0


This code displays 10 random records [due to .sample(10) method] from the login_timestamp and Login_Hour columns to verify that the Login_Hour column has been created correctly.

###Step 6 – Logout_Hour

In [37]:
clean_df["Logout_Hour"] = clean_df["logout_timestamp"].dt.hour

This line extracts the hour from the logout_timestamp column and stores it in a new column called Logout_Hour.

Verify:

In [38]:
clean_df[
    ["logout_timestamp", "Logout_Hour"]
].sample(10)

,logout_timestamp,Logout_Hour
21104,2026-02-04 17:34:48,17.0
29142,2026-02-18 17:11:59,17.0
39864,2026-03-06 19:36:15,19.0
2302,2026-01-06 14:42:45,14.0
4021,2026-01-08 18:58:23,18.0
18673,2026-02-02 19:36:48,19.0
14419,2026-01-23 23:06:12,23.0
17193,2026-01-29 18:29:53,18.0
31192,2026-02-20 18:39:18,18.0
33219,2026-02-24 23:57:30,23.0


This code displays 10 random records of the logout_timestamp and Logout_Hour columns to verify that the Logout_Hour has been extracted correctly from the logout_timestamp.

###Step 7 — Late_Arrival_Flag

It tells us about how many attendance records involve a late arrival.

In [39]:
clean_df["Late_Arrival_Flag"] = clean_df["late_arrival_mins"].apply(
    lambda x: "Late" if x > 0 else "On Time"
)

The **.apply()** method uses a **lambda (anonymous) function**, where **x** represents each value in the **late_arrival_mins** column, to assign **Late** if x > 0; otherwise, it assigns **On Time** and stores the result in the **Late_Arrival_Flag** column.

Verify:

In [40]:
clean_df[
    ["late_arrival_mins", "Late_Arrival_Flag"]
].sample(10)

,late_arrival_mins,Late_Arrival_Flag
52450,25,Late
46238,0,On Time
21717,0,On Time
35337,0,On Time
31096,0,On Time
19909,0,On Time
9506,0,On Time
2216,0,On Time
41431,3,Late
5370,49,Late


The clean_df DataFrame is used to select the **late_arrival_mins** and **Late_Arrival_Flag** columns, and the .sample(10) function displays 10 random records to verify that the Late_Arrival_Flag has been created correctly.

In [41]:
clean_df["Late_Arrival_Flag"].value_counts()

,count
Late_Arrival_Flag,
On Time,27891
Late,27483


This code counts the number of employees who were "On Time" and the number of employees who were "Late".

###Step 8 — Overtime_Flag

It tells us whether the employee work overtime on that attendance day or not

In [42]:
clean_df["Overtime_Flag"] = clean_df["overtime_hours"].apply(
    lambda x: "Yes" if x > 0 else "No"
)

The **.apply()** method uses a **lambda (anonymous) function**, where **x** represents each value in the **overtime_hours** column, to label the record as **Yes** if x > 0; otherwise, it labels it as **No** and stores the result in the **Overtime_Flag** column

Verify:

1. Random sample

In [43]:
clean_df[
    ["overtime_hours", "Overtime_Flag"]
].sample(10)

,overtime_hours,Overtime_Flag
41373,0.00,No
2492,0.00,No
24513,0.47,Yes
47068,0.59,Yes
12418,0.16,Yes
17900,0.00,No
43334,0.08,Yes
8421,0.00,No
2813,0.00,No
16289,0.00,No


This code selects the **overtime_hours** and **Overtime_Flag** columns from the **clean_df** DataFrame. The .sample(10) method is then used to display 10 random records so that the **Overtime_Flag** values can be verified against the corresponding **overtime_hours** values.

2. Count

In [44]:
clean_df["Overtime_Flag"].value_counts()

,count
Overtime_Flag,
No,42232
Yes,13142


- This code counts how many employees worked overtime (Yes) and how many did not work overtime (No).
- clean_df : DataFrame
- ["Overtime_Flag"] : Selects the Overtime_Flag column containing 2 values Yes, No
- The value_counts() method counts how many times each unique value appears in the selected column.

###Step 9 — Productivity_Category

It tells us how can we classify employees based on their productive working hours

In [45]:
def productivity_category(hours):
    if hours >= 8:
        return "Highly Productive"
    elif hours >= 6:
        return "Moderately Productive"
    else:
        return "Low Productive"

clean_df["Productivity_Category"] = clean_df["net_productive_hours"].apply(productivity_category)

A custom function named **productivity_category()** is created using **def**. The function receives one value **(hours)** at a time from the **net_productive_hours** column. It checks the **productive hours** using **if, elif, and else** conditions. If the hours are 8 or more, it returns Highly Productive; if the hours are between 6 and less than 8, it returns Moderately Productive; otherwise, it returns Low Productive. Finally, the **.apply()** method applies this function to every value in the **net_productive_hours** column, and the results are stored in a new column called **Productivity_Category**.


Verify:

Random Sample

In [46]:
clean_df[
    ["net_productive_hours", "Productivity_Category"]
].sample(10)

,net_productive_hours,Productivity_Category
30905,8.20,Highly Productive
19837,8.05,Highly Productive
39456,10.92,Highly Productive
42222,0.00,Low Productive
6822,8.54,Highly Productive
36775,8.98,Highly Productive
24780,8.92,Highly Productive
20421,8.55,Highly Productive
31708,8.26,Highly Productive
598,7.55,Moderately Productive


- clean_df refers to the cleaned DataFrame
- Two columns(net_productive_hours,Productivity_Category) are selected from clean_df
- .sample(10) method selects randomly 10 rows from the selected columns

Count

In [47]:
clean_df["Productivity_Category"].value_counts()

,count
Productivity_Category,
Highly Productive,31761
Moderately Productive,17571
Low Productive,6042


This code counts how many employees belong to each productivity category: Highly Productive, Moderately Productive, and Low Productive.

###Final Validation





1. Check Final Shape

In [48]:
print("Final Shape:", clean_df.shape)

Final Shape: (55374, 31)


2. Check Dataset Information

In [49]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55374 entries, 0 to 55373
Data columns (total 31 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   attendance_id          55374 non-null  object        
 1   employee_id            55374 non-null  object        
 2   employee_name          55374 non-null  object        
 3   gender                 55374 non-null  object        
 4   department             55374 non-null  object        
 5   designation            55374 non-null  object        
 6   employment_type        55374 non-null  object        
 7   office_location        55374 non-null  object        
 8   date_of_joining        55374 non-null  datetime64[ns]
 9   attendance_date        55374 non-null  datetime64[ns]
 10  shift_type             55374 non-null  object        
 11  attendance_status      55374 non-null  object        
 12  work_mode              55374 non-null  object        
 13  l

3. Check Missing Values

In [50]:
clean_df.isnull().sum()

,0
attendance_id,0
employee_id,0
employee_name,0
gender,0
department,0
designation,0
employment_type,0
office_location,0
date_of_joining,0
attendance_date,0


Expected:

Most columns → 0<br/>
But,
login_timestamp
logout_timestamp
Login_Hour
Logout_Hour

may still have missing values because employees on leave or absent do not have login/logout records. That is expected and should remain as-is.

4. Check Duplicate Rows

In [51]:
clean_df.duplicated().sum()

np.int64(0)

5. Preview the Final Dataset

In [52]:
clean_df.head()

,attendance_id,employee_id,employee_name,gender,department,designation,employment_type,office_location,date_of_joining,attendance_date,...,leave_type,Joining_Year,Employee_Tenure_Years,Attendance_Month,Attendance_Weekday,Login_Hour,Logout_Hour,Late_Arrival_Flag,Overtime_Flag,Productivity_Category
0,ATT0000001,VL1000,Arjun Verma,Male,Engineering,Software Engineer Intern,Intern,Koramangala HQ,2025-06-26,2026-01-02,...,Not Applicable,2025,1,January,Friday,9.0,18.0,Late,No,Moderately Productive
1,ATT0000002,VL1001,Rekha Chatterjee,Female,Operations,Operations Manager,Full-Time,Koramangala HQ,2024-02-16,2026-01-02,...,Not Applicable,2024,2,January,Friday,9.0,18.0,On Time,No,Highly Productive
2,ATT0000003,VL1002,Tanmay Bhardwaj,Male,Operations,Operations Analyst,Full-Time,Whitefield Tech Park,2020-10-13,2026-01-02,...,Not Applicable,2020,6,January,Friday,9.0,20.0,Late,Yes,Highly Productive
3,ATT0000004,VL1003,Aditya Bose,Female,Customer Success,Customer Success Executive,Full-Time,Koramangala HQ,2022-01-10,2026-01-02,...,Not Applicable,2022,4,January,Friday,13.0,22.0,On Time,No,Moderately Productive
4,ATT0000005,VL1004,Uday Acharya,Male,Human Resources,Talent Acquisition Partner,Full-Time,Electronic City Campus,2024-09-09,2026-01-02,...,Not Applicable,2024,2,January,Friday,10.0,20.0,Late,No,Highly Productive


In [53]:
clean_df['attendance_date'].head(10)

,attendance_date
0,2026-01-02
1,2026-01-02
2,2026-01-02
3,2026-01-02
4,2026-01-02
5,2026-01-02
6,2026-01-02
7,2026-01-02
8,2026-01-02
9,2026-01-02


In [54]:
print(df.dtypes['attendance_date'])

object


In [55]:
print(clean_df.dtypes['attendance_date'])

datetime64[ns]


In [56]:
print(df['attendance_date'].head())

0    2026-01-02
1    2026-01-02
2    2026-01-02
3    2026-01-02
4    2026-01-02
Name: attendance_date, dtype: object


In [57]:
print(clean_df['attendance_date'].head())

0   2026-01-02
1   2026-01-02
2   2026-01-02
3   2026-01-02
4   2026-01-02
Name: attendance_date, dtype: datetime64[ns]


In [58]:
clean_df.to_csv(
    'employee_attendance_bangalore_q1_2026_cleaned_new.csv',
    index=False,
    date_format='%Y-%m-%d'
)

In [59]:
clean_df['Late_Arrival_Flag'].value_counts()

,count
Late_Arrival_Flag,
On Time,27891
Late,27483


In [60]:
clean_df[['late_arrival_mins', 'Late_Arrival_Flag']].head(20)

,late_arrival_mins,Late_Arrival_Flag
0,12,Late
1,0,On Time
2,15,Late
3,0,On Time
4,20,Late
5,1,Late
6,7,Late
7,24,Late
8,17,Late
9,11,Late


In [61]:
clean_df["is_late"] = clean_df["Late_Arrival_Flag"].map({
    "Late": 1,
    "On Time": 0
})

The .map() method is used to replace or convert the values of a column according to a mapping provided in a dictionary. In this code, it converts the text values 'Late' and 'On Time' into numeric values 1 and 0, and stores the result in a new column called is_late

In [62]:
clean_df[["Late_Arrival_Flag", "is_late"]].head(10)

,Late_Arrival_Flag,is_late
0,Late,1
1,On Time,0
2,Late,1
3,On Time,0
4,Late,1
5,Late,1
6,Late,1
7,Late,1
8,Late,1
9,Late,1


In [63]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new2.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [64]:
clean_df["is_early_exit"] = (clean_df["early_exit_mins"] > 0).astype(int)

- A new column called is_early_exit is created to indicate whether an employee left early.
- .astype() = change the data type
- clean_df["early_exit_mins"] > 0 produces Boolean values:
  - True → employee left early
  - False → employee did not leave early
- .astype(int) converts those Boolean values into integers:
    - True → 1
    - False → 0

In [65]:
clean_df[["early_exit_mins", "is_early_exit"]].head(10)

,early_exit_mins,is_early_exit
0,0,0
1,6,1
2,0,0
3,36,1
4,0,0
5,0,0
6,0,0
7,0,0
8,0,0
9,52,1


In [66]:
clean_df["is_early_exit"].value_counts(normalize=True) * 100

,proportion
is_early_exit,
0,67.605736
1,32.394264


This code calculates the percentage of attendance records where employees did and did not leave early.

In [67]:
clean_df["is_early_exit"]

,is_early_exit
0,0
1,1
2,0
3,1
4,0
...,...
55369,0
55370,0
55371,0
55372,0


This code selects and displays the is_early_exit column from the clean_df DataFrame.

In [68]:
clean_df["is_early_exit"].mean() * 100

np.float64(32.39426445624301)

This calculates the percentage of attendance records where employees left early.

In [69]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_final_new3.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [70]:
clean_df["overtime_hours_paid"] = clean_df["overtime_hours"].where(
    clean_df["overtime_hours"] > 0
)

Keeps overtime hours only when they are greater than 0; otherwise, NaN is stored.

In [71]:
clean_df[["overtime_hours", "overtime_hours_paid"]].head(10)

,overtime_hours,overtime_hours_paid
0,0.00,NaN
1,0.00,NaN
2,0.58,0.58
3,0.00,NaN
4,0.00,NaN
5,0.00,NaN
6,0.00,NaN
7,0.00,NaN
8,0.00,NaN
9,0.00,NaN


Displays the first 10 records to check whether the new column (overtime_hours_paid) was created correctly.

In [72]:
clean_df["overtime_hours_paid"].describe()

,overtime_hours_paid
count,13142.000000
mean,0.632232
std,0.513294
min,0.010000
25%,0.230000
50%,0.510000
75%,0.910000
max,3.860000


Provides statistics such as count, mean, standard deviation, minimum, quartiles, and maximum for the paid overtime hours.

In [73]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new4.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [74]:
print(clean_df.columns.tolist())

['attendance_id', 'employee_id', 'employee_name', 'gender', 'department', 'designation', 'employment_type', 'office_location', 'date_of_joining', 'attendance_date', 'shift_type', 'attendance_status', 'work_mode', 'login_timestamp', 'logout_timestamp', 'total_hours_worked', 'break_duration_mins', 'net_productive_hours', 'late_arrival_mins', 'early_exit_mins', 'overtime_hours', 'leave_type', 'Joining_Year', 'Employee_Tenure_Years', 'Attendance_Month', 'Attendance_Weekday', 'Login_Hour', 'Logout_Hour', 'Late_Arrival_Flag', 'Overtime_Flag', 'Productivity_Category', 'is_late', 'is_early_exit', 'overtime_hours_paid']


- This code displays the names of all columns present in the clean_df DataFrame as a Python list.
- clean_df : clean dataframe
- .columns : Gets the column names of the DataFrame.
- .tolist() : Converts the column names into a normal Python list.
- print() : Displays the resulting list as output.

In [75]:
[col for col in clean_df.columns if "productive" in col.lower() or "hours" in col.lower()]

['total_hours_worked',
 'net_productive_hours',
 'overtime_hours',
 'overtime_hours_paid']

- Finds column names containing "productive" or "hours".


In [76]:
clean_df["productivity_ratio"] = (
    clean_df["net_productive_hours"] /
    clean_df["total_hours_worked"]
)

- Creates the new productivity_ratio column by calculating:
- Net Productive Hours ÷ Total Hours Worked

In [77]:
clean_df["productivity_ratio"].head(10)

,productivity_ratio
0,0.838782
1,0.879180
2,0.909782
3,0.852154
4,0.894954
5,0.831183
6,0.855397
7,0.905956
8,0.864286
9,0.903266


Displays the first 10 values to check whether the ratio was calculated correctly.

In [78]:
clean_df["productivity_ratio"].describe()

,productivity_ratio
count,52739.000000
mean,0.883420
std,0.044008
min,0.472222
25%,0.860664
50%,0.886410
75%,0.911487
max,1.000000


- Gives the statistical summary: count, mean, standard deviation, minimum, quartiles, and maximum.
- .describe() tells us how the data is spread

In [79]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new5.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [80]:
clean_df["work_mode"].value_counts()

,count
work_mode,
Work From Office,40283
Work From Home,10906
Not Applicable,2635
Client Site,1550


Counts the number of attendance records for each work mode.

In [81]:
clean_df["work_mode"].value_counts(normalize=True) * 100

,proportion
work_mode,
Work From Office,72.747138
Work From Home,19.695164
Not Applicable,4.758551
Client Site,2.799148


- The code calculates the percentage of attendance records for each work mode.
- .value_counts() : Counts how many times each value appears.
- normalize=True : Instead of returning the actual count, it returns the proportion of the total.
- * 100 : Converts the proportion into a percentage.

In [82]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new6.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [83]:
clean_df["is_on_leave"] = clean_df["attendance_status"].map({
    "On Leave": 1
}).fillna(0).astype(int)

- We created a new binary column called is_on_leave to identify whether an employee was on leave or not.
- Used .map() : Converted "On Leave" → 1.
- .fillna(0) : All other statuses that became NaN were converted to 0
- .astype(int) : Converted the results into integer values: 0 and 1.

In [84]:
clean_df["is_on_leave"].value_counts()

,count
is_on_leave,
0,52739
1,2635


We are counting how many records have each value (0 and 1) in the is_on_leave column.

In [85]:
clean_df[["attendance_status", "is_on_leave"]].head(15)

,attendance_status,is_on_leave
0,Present,0
1,Present,0
2,Present,0
3,Present,0
4,Present,0
5,Present,0
6,Present,0
7,Present,0
8,Half Day,0
9,Present,0


In [86]:
clean_df["is_on_leave"] = clean_df["attendance_status"].map({
    "On Leave": 1
}).fillna(0).astype(int)

clean_df["is_on_leave"].value_counts()

,count
is_on_leave,
0,52739
1,2635


- The existing attendance_status column is checked.
- "On Leave" is converted to 1 using .map().
- All other statuses become NaN, which are then converted to 0 using .fillna(0).
- .astype(int) converts the values to integers.
- The result is stored in a new column called is_on_leave.

In [87]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new7.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [88]:
print(clean_df.columns.tolist())

['attendance_id', 'employee_id', 'employee_name', 'gender', 'department', 'designation', 'employment_type', 'office_location', 'date_of_joining', 'attendance_date', 'shift_type', 'attendance_status', 'work_mode', 'login_timestamp', 'logout_timestamp', 'total_hours_worked', 'break_duration_mins', 'net_productive_hours', 'late_arrival_mins', 'early_exit_mins', 'overtime_hours', 'leave_type', 'Joining_Year', 'Employee_Tenure_Years', 'Attendance_Month', 'Attendance_Weekday', 'Login_Hour', 'Logout_Hour', 'Late_Arrival_Flag', 'Overtime_Flag', 'Productivity_Category', 'is_late', 'is_early_exit', 'overtime_hours_paid', 'productivity_ratio', 'is_on_leave']


This is used to display the complete list of column names currently present in clean_df.

In [89]:
[col for col in clean_df.columns if "Attendance" in col or "attendance" in col]

['attendance_id',
 'attendance_date',
 'attendance_status',
 'Attendance_Month',
 'Attendance_Weekday']

- This is used to find all column names in clean_df that contain either Attendance or attendance.
- clean_df.columns
→ Gets all column names from the DataFrame.
- for col in clean_df.columns
→ Checks each column name one by one.
Here, col represents the current column name.
- "Attendance" in col
→ Checks whether the column name contains Attendance with a capital A.
- "attendance" in col
→ Checks whether the column name contains attendance with a lowercase a.
or either condition can be true.
- If either condition is true, that column name is added to the resulting list.

In [90]:
clean_df["Attendance_Flag"] = clean_df["attendance_status"].map({
    "Present": 1,
    "Half Day": 1,
    "On Leave": 0
}).fillna(0).astype(int)

- This code creates a new Attendance_Flag column from the attendance_status column. Present and Half Day are mapped to 1, while On Leave is mapped to 0. Any other or missing status is also converted to 0 using .fillna(0), and .astype(int) converts the final values into integers. Therefore, 1 represents an attendance day and 0 represents a non-attendance/leave record.

In [91]:
clean_df[["attendance_status", "Attendance_Flag"]].head(10)

,attendance_status,Attendance_Flag
0,Present,1
1,Present,1
2,Present,1
3,Present,1
4,Present,1
5,Present,1
6,Present,1
7,Present,1
8,Half Day,1
9,Present,1


In [92]:
print(clean_df.columns.tolist())

['attendance_id', 'employee_id', 'employee_name', 'gender', 'department', 'designation', 'employment_type', 'office_location', 'date_of_joining', 'attendance_date', 'shift_type', 'attendance_status', 'work_mode', 'login_timestamp', 'logout_timestamp', 'total_hours_worked', 'break_duration_mins', 'net_productive_hours', 'late_arrival_mins', 'early_exit_mins', 'overtime_hours', 'leave_type', 'Joining_Year', 'Employee_Tenure_Years', 'Attendance_Month', 'Attendance_Weekday', 'Login_Hour', 'Logout_Hour', 'Late_Arrival_Flag', 'Overtime_Flag', 'Productivity_Category', 'is_late', 'is_early_exit', 'overtime_hours_paid', 'productivity_ratio', 'is_on_leave', 'Attendance_Flag']


In [93]:
clean_df.to_csv(
    "employee_attendance_bangalore_q1_2026_cleaned_new8.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [94]:
clean_df["tenure_days"] = (
    clean_df["attendance_date"] - clean_df["date_of_joining"]
).dt.days

- attendance_date = date of the attendance record.
- date_of_joining = date when the employee joined.
- Both dates are subtracted to find the tenure_days.
- .dt.days converts the resulting time difference into number of days.
- The result is stored in a new column called tenure_days.

In [95]:
daily_attendance_rate = clean_df.groupby("attendance_date")["Attendance_Flag"].mean().reset_index()

- groupby("attendance_date") groups all records according to their attendance date.
- ["Attendance_Flag"] selects the attendance indicator.
- .mean() calculates the average Attendance_Flag for each date.
- .reset_index() converts the grouped result back into a normal DataFrame.
- The result is stored in daily_attendance_rate.

In [96]:
clean_df["tenure_days"] = (
    clean_df["attendance_date"] - clean_df["date_of_joining"]
).dt.days

- clean_df["attendance_date"]
→ Selects the employee's attendance date.

- clean_df["date_of_joining"]
→ Selects the employee's joining date.

- attendance_date - date_of_joining
→ Calculates the difference between the two dates.

In [97]:
clean_df["tenure_days"].describe()

,tenure_days
count,55374.000000
mean,1027.294181
std,634.468926
min,0.000000
25%,465.000000
50%,970.500000
75%,1592.000000
max,2265.000000


In [98]:
clean_df["attendance_month_num"] = clean_df["attendance_date"].dt.month

- creates a new column called attendance_month_num that contains the month number extracted from attendance_date.
- clean_df["attendance_date"]
→ Selects the attendance_date column.
- .dt
→ Allows us to perform date/time operations because attendance_date is a datetime column.
- .month
→ Extracts the month number from each date:
  - January → 1
  - February → 2
  - March → 3
  - December → 12
- clean_df["attendance_month_num"]
→ Stores these month numbers in a new column named attendance_month_num

In [99]:
clean_df["Quarter"] = "Q1 2026"

It creates a new column called Quarter and assigns the value "Q1 (January–March) of 2026" to every row in the DataFrame.

In [100]:
clean_df["Attendance_Year"] = clean_df["attendance_date"].dt.year

- Means a new column called Attendance_Year is created by extracting the year from the attendance_date column.
- clean_df["attendance_date"]
→ Selects the attendance_date column.
- .dt → Allows date/time information to be extracted from the datetime column.
- .year → Extracts only the year from each date.

In [101]:
clean_df["attendance_month_num"] = clean_df["attendance_date"].dt.month
clean_df["Attendance_Year"] = clean_df["attendance_date"].dt.year
clean_df["Quarter"] = "Q1 2026"

- Create attendance_month_num
   - Takes the attendance_date.
   - Extracts the month number using .dt.month.
   - Stores it in a new column called attendance_month_num
     - January → 1
     - February → 2
     - March → 3
- Create Attendance_Year
   - Takes the attendance_date.
   - Extracts the year using .dt.year.
   - Stores it in a new column called Attendance_Year
- Create Quarter
   - Creates a new column called Quarter
   - Assigns "Q1 2026" to every row.
   - It does not calculate the quarter from the date.
Since your dataset covers January–March 2026, all records belong to Q1 2026.

In [102]:
clean_df.to_csv(
    "employee_attendance_dashboard_cleaned_new9.csv",
    index=False,
    date_format="%Y-%m-%d"
)

In [103]:
clean_df["work_mode"].value_counts(dropna=False)

,count
work_mode,
Work From Office,40283
Work From Home,10906
Not Applicable,2635
Client Site,1550


Count how many times each different value appears in the work_mode column, including any missing (NaN) values.
- clean_df["work_mode"]
→ Selects the work_mode column.
- .value_counts()
→ Counts the number of records for each unique work mode.
- dropna=False
→ Tells Pandas to also count missing values (NaN) instead of ignoring them.

In [104]:
clean_df["leave_type"].value_counts(dropna=False)

,count
leave_type,
Not Applicable,52739
Casual Leave,796
Sick Leave,713
Earned Leave,604
Loss of Pay,166
Work From Home Comp-Off,141
Parental Leave,109
Bereavement Leave,57
Marriage Leave,49


- It counts how many times each type of leave appears in the leave_type column, including missing values.
- clean_df["leave_type"]
→ Selects the leave_type column from the DataFrame.
- .value_counts()
→ Counts how many records belong to each unique leave type.
- dropna=False
→ Makes sure that missing values (NaN) are also counted.

In [105]:
clean_df.groupby("shift_type")["late_arrival_mins"].describe()

,count,mean,std,min,25%,50%,75%,max
shift_type,,,,,,,,
Early (08:00-17:00),4830.0,8.964803,19.633271,0.0,0.0,0.0,12.0,179.0
Flexible (10:30-19:30),11130.0,10.722013,20.025404,0.0,0.0,2.0,15.0,190.0
General (09:30-18:30),23919.0,9.643463,19.018901,0.0,0.0,0.0,13.0,218.0
Late/US Overlap (14:00-23:00),7430.0,9.726783,19.468145,0.0,0.0,0.0,13.0,187.0
Mid (11:00-20:00),8065.0,9.960942,19.843649,0.0,0.0,0.0,13.0,181.0


- It gives a statistical summary of late-arrival minutes separately for each shift type.
- clean_df.groupby("shift_type") : Groups the attendance records according to shift_type.
- ["late_arrival_mins"] : → Selects the late_arrival_mins column from each shift group.
- .describe() : Calculates statistics for late-arrival minutes for each shift separately, such as:
     - count → number of records
     - mean → average late-arrival minutes
     - std → variation in late-arrival minutes
     - min → minimum late-arrival minutes
     - 25%, 50%, 75% → quartiles
     - max → maximum late-arrival minutes

In [106]:
clean_df.groupby("designation")["net_productive_hours"].describe()

,count,mean,std,min,25%,50%,75%,max
designation,,,,,,,,
Account Executive,2736.0,7.912595,2.124948,0.0,7.6200,8.350,9.0400,12.14
Accounts Manager,570.0,7.446614,2.014969,0.0,7.0925,7.835,8.5100,10.55
Analytics Engineer,684.0,7.767412,2.182121,0.0,7.5000,8.165,8.9025,11.74
Associate Product Manager,1026.0,7.745565,2.416286,0.0,7.5125,8.410,9.0575,11.30
Associate Software Engineer,3876.0,8.026713,2.179878,0.0,7.6475,8.460,9.2300,12.71
Content Strategist,684.0,7.750629,2.112335,0.0,7.4600,8.160,8.8125,11.79
Customer Success Executive,1596.0,7.623083,2.205687,0.0,7.2075,8.050,8.8400,12.86
Customer Success Manager,1425.0,7.890575,2.044797,0.0,7.5000,8.270,9.0000,11.79
Data Analyst,1225.0,7.200408,2.592909,0.0,6.4600,7.990,8.9100,11.63


- It gives a statistical summary of net_productive_hours for each employee designation.
- clean_df.groupby("designation")
→ Groups all attendance records according to employee designation.
- ["net_productive_hours"]
→ Selects the net_productive_hours column for each designation.
.describe()
→ Calculates statistics for productive hours within each designation:
    - count → number of records
    - mean → average productive hours
    - std → variation in productive hours
    - min → minimum productive hours
    - 25% → first quartile
    - 50% → median
    - 75% → third quartile
    - max → maximum productive hours

In [107]:
clean_df[["Attendance_Weekday", "department", "Late_Arrival_Flag"]].head()

,Attendance_Weekday,department,Late_Arrival_Flag
0,Friday,Engineering,Late
1,Friday,Operations,On Time
2,Friday,Operations,Late
3,Friday,Customer Success,On Time
4,Friday,Human Resources,Late


- clean_df
→ Refers to your cleaned DataFrame.
- [["Attendance_Weekday", "department", "Late_Arrival_Flag"]]
→ Selects these 3 columns:
- Attendance_Weekday → day of the week, such as Monday, Tuesday, etc.
- department → employee's department.
- Late_Arrival_Flag → indicates whether the employee arrived late (Late) or on time (On Time).
- .head()
→ Displays the first 5 rows by default.

#End of Cleaning Notebook